File: common/src/main/java/dev/obscuria/fragmentum/ModCompat.java
package dev.obscuria.fragmentum;

import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public record ModCompat(String modId, String modName) {

    public ModCompat(String modId) {
        this(modId, modId);
    }

    public String ModName() {
        return modName;
    }

    public boolean isLoaded() {
        return Fragmentum.PLATFORM.isModLoaded(modId);
    }

    public <T> @Nullable T getIfLoaded(Supplier<Supplier<T>> supplier) {
        if (!isLoaded()) return null;
        return supplier.get().get();
    }

    public void runIfLoaded(Supplier<Runnable> runnable) {
        if (!isLoaded()) return;
        runnable.get().run();
    }

    public void runIfMissing(Runnable runnable) {
        if (isLoaded()) return;
        runnable.run();
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/ModCompat.kt
@file:Suppress("unused", "RedundantNullableReturnType")

package dev.obscuria.fragmentum

class ModCompat(
    val modId: String,
    val displayName: String
) {

    val isLoaded: Boolean get() = Fragmentum.PLATFORM.isModLoaded(modId)

    constructor(modId: String) : this(modId, modId)

    inline fun <T> getIfLoaded(supplier: () -> () -> T): T? {
        if (isLoaded) return supplier()()
        return null
    }

    inline fun runIfLoaded(runnable: () -> () -> Any) {
        if (!isLoaded) return
        runnable()()
    }

    inline fun runIfMissing(runnable: () -> Any) {
        if (isLoaded) return
        runnable()
    }
}